File: python/consumer.py
#!/usr/bin/env python
import sys

from amqp import EXCHANGE_NAME
from amqp import connect_get_channel_declare_exchange_and_return_channel


if __name__ == "__main__":
    DEBUG = False
    ROUTING_KEY = "*.info" if len(sys.argv) < 2 else sys.argv[1]
    QUEUE_NAME = "" if len(sys.argv) < 3 else sys.argv[2]

    channel = connect_get_channel_declare_exchange_and_return_channel()

    queue = channel.queue_declare(
        queue=QUEUE_NAME,
        auto_delete=True,
    )
    QUEUE_NAME = queue.method.queue

    channel.queue_bind(
        queue=QUEUE_NAME,
        exchange=EXCHANGE_NAME,
        routing_key=ROUTING_KEY
    )

    print("Waiting for messages routed to '%s' with binding-key '%s' from '%s'. "
          "To exit press CTRL+C" % (QUEUE_NAME, ROUTING_KEY, EXCHANGE_NAME))


    def callback(ch, method, properties, body):
        print("-" * 80)
        if DEBUG:
            print("     Channel: %s" % ch)
            print("     Method: %s" % method)
            print("     Properties: %s" % properties)
            print("")
        print("     Routing-Key: '%s'" % method.routing_key)
        print("     Body %r" % body)


    channel.basic_consume(
        queue=QUEUE_NAME,
        auto_ack=True,
        on_message_callback=callback
    )

    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        pass

File: go/consumer.go
package main

import (
	"github.com/streadway/amqp"
	"log"
	"os"
)


const AmqpUri = "amqp://guest:guest@localhost:5672/"
const ExchangeName = "amqp-training"


func argAtIndexOrDefault(index int, defaultValue string) string {
	result := defaultValue

	if len(os.Args) > index + 1 {
		result = os.Args[index]
	}

	return result
}


func main() {
	RoutingKey := argAtIndexOrDefault(1, "#")
	QueueName := argAtIndexOrDefault(2, "go-consumer")

	connection, err := amqp.Dial(AmqpUri)
	if err != nil {
		log.Fatalf("Failed to connect to %s: %s", AmqpUri, err)
	}
	defer connection.Close()
	log.Printf("Connected to AMQP: %s", AmqpUri)


	channel, err := connection.Channel()
	if err != nil {
		log.Fatalf("Failed to acquire channel: %s", err)
	}
	defer channel.Close()


	err = channel.ExchangeDeclare(ExchangeName,"topic",false,false,false,false,nil,)
	if err != nil {
		log.Fatalf("Failed to declare exchange '%s': %s", ExchangeName, err)
	}

	_, err = channel.QueueDeclare(QueueName, false, true, false, false, nil, )
	if err != nil {
		log.Fatalf("Failed to declare queue '%s': %s", QueueName, err)
	}
	log.Printf("Using queue: %s", QueueName)


	err = channel.QueueBind(QueueName, RoutingKey, ExchangeName, false, nil)
	if err != nil {
		log.Fatalf("Failed to bind queue '%s' to exchange '%s' with '%s': %s", QueueName, ExchangeName, RoutingKey, err)
	}
	log.Printf("Bound queue '%s' to exchange '%s' with '%s'", QueueName, ExchangeName, RoutingKey)


	messages, err := channel.Consume(QueueName, "",true, false, false, false, nil)

	if err != nil {
		log.Printf("Failed to register a consumer: %s", err)
	}


	go func() {
		for d := range messages {
			log.Printf(" [x] %24s:  %s", d.RoutingKey, d.Body)
		}
	}()

	log.Printf(" [*] Waiting for messages. To exit press CTRL+C")

	forever := make(chan bool)
	<-forever
}
